### ISSUE TITLE
Implement quantile regression and other non-shadow attacks

### ISSUE THREAD

**Issue by jim-smith** [AI-SDC/SACRO-ML#306]
respectively
https://proceedings.neurips.cc/paper_files/paper/2023/file/01328d0767830e73a612f9073e9ff15f-Paper-Conference.pdf
code at 
https://github.com/amazon-science/quantile-mia

and
https://arxiv.org/pdf/2312.03262
code at
https://github.com/privacytrustlab/ml_privacy_meter/tree/master/research/

**Comment by jim-smith** [comment]
and others that don't use shadow models

**Comment by jim-smith** [comment]
@albacrespi @jamesliley I think this fits into is part the body of work following on from looking at distance to boundaries?

**Comment by jim-smith** [comment]
# Outline sketch of explainer for Shamy and Hasaan

## Background 
Most membership inference attacks work on the basis of observing something about how a target model $f$ behaves in response to a given data point/record $x$. This could be the error, the loss (can be slightly different) or other things.  
- in general we will call this the _test statistic_ and this clearly depends on the model and the data point.

### Confidence Scores
Lots of attacks are based on the model's _confidence_ score.
- assume our data is one-hot encoded and has $C$ different class labels 
- so for an input $x$ with true label $y$ , the model $f$    
   (for example,  through a  function called ```predict_proba``` in scikit-learn)   
   outputs a vector $f(x)= \hat{y}$ of $C$ values (each in the range $[0,1]$)   
  the components of this vector  represent how strongly it thinks $x$ belongs to each class.
    - I.e. $\hat{y}_1$ is the strength of $f$'s belief that $y== class1$ 

- In the simplest version we just say that the model's confidence for the record-label pair $(x,y)$ is given by $\hat{y}_i$ where $i==y$
     - This simple version is not great though, because it doesn't take into account that even if the model is only little confident about other $C-1$ classes, then the size of $\hat{y}_I$ inevitably goes down as $C$ increases
 - So a more sophisticated version compares the model's confidence for the true label to its highest confidence for any other label.
    -  I.e. if the prediction vector $f(x) = [0.2, 0.5,0.4,0.1]$ for a 4 class problem (classes (_A,B,C,D_) one-hot encoded),
    -  and the true class $y$ for input $x$ is _B_ 
    - then the model's $confidence(f, x, y) = 0.5 - 0.4 =0.1$  (0.4 is the highest value of the wrong classes)

- ASIDE: 
    -  If we divide each of these $\hat{y}$ value by their total (I.e. normalise) then these values sum to 1.0 and can be considered to be the prediction probabilities.
    - However neither quantile regression nor the Likelihood ratio attack do this for neural models.  
       Instead they work on what they call the _logits_ :  the outputs from the $C$ final layer nodes   _before_ softmax is applied. (because this gives a cleaner signal)
    -   In the equations in the  Quantile regression paper these values are denoted by $z$ 
    - So the description above is formally described by their equation $s(x,y) = z_y(x) − max_{y′ \neq y} z_{y′ }(x)$

## Simplified versions of some attacks

### Simple loss-based attacks (Yeom 2018)
Given a trained model $f$ and a set of records $\mathcal{D}^{out}$ that we know were not used for training.

1. For each pair $(x',y') \in \mathcal{D}^{out}$ measure the $loss_f(x',y')$ 
2. Make a histogram of the values and fit a curve to it 
   so on a graph it would have  test static $loss(_f(x',y')$  on x-axis,   P(observing that value) on y-axis
3. For the record $(x,y)$ I think _might_ be in the training set
    3.1 Measure $loss_f(x,y)$
     3.2 See where it is on the x-axis - and measure the height of the curve there - call this $\tau$
     3.3 Use that as an estimate of whether the model behaved for $(x,y)$ like it did for other non-training points
      3.4  If curve is quite high at the value $loss_f(x,y)$ that means the target model $f$ recorded similar losses for other _unseen_ records from $\mathcal{D}^{out}$.
      3.5 But if the curve is really low at the value $loss_f(x,y)$ that means the model didn't make that size loss for (m)any of the members of $\mathcal{D}^{out}$   - so the record is likely in the training set
 4. Membership inference attack is basically just step 3, then predict 'training set' if $\tau >t$ for some fixed threshold $t

<img width="895" height="380" alt="Image" src="https://github.com/user-attachments/assets/79b87a87-3e6c-447f-a221-009a3f71f348" />

**this is roughly what our worst-case** attacks do

**Problem 1** this doesn't take into account how difficult $(x,y)$ was to predict

**Problem 2** this depends a lot on the set of non-training records $\mathcal{D}^{out}$ - which introduces lots of variations


## (crudely speaking) Shadow-model based attacks
Build on this idea and also try to take account of uncertainty / noise effects caused by training - for example specific random weights when a neural network is initialised.

1. Take your training set $\mathcal{D}^{in}$  and non-training set $\mathcal{D}^{out}$ 
- Mix them up and then do lots of random train/test splits
- train a 'shadow model' - same config as the original- on each split
- Goal is to train $n$ models so  that each record will be:
      -    in the training set of $\approx n/2$ of them (call these the _in models_ for that record) a
      - in the test set for the other $\approx n/2$ (call these the _out models_ for that record).
- record your choice of test statistic  for every record for each of these different models


2. Given a new record $(x,y)$ I want to predict if it was in the training set.
- measure $S_f(x,y)$ for my target model $f$
- take the $n/2$ 'out' models and measure the test statistic for each of them
    - plot these on a histogram and generate a probability curve like above
    - then I can look up the height of the curve for my record $S_f(x,y)$ to say :
         - _How much is my target model $f$ behaving like a model that was **not** trained using $(x,y)$ _     
- do the same for the _out models_ to answer the question
       - _How much is my target model $f$ behaving like a model that was **not** trained using $(x,y)$ _     

- Depending on the specific attack, predict with $(x,y)$ was used for training according to where it meets those two curves

<img width="888" height="413" alt="Image" src="https://github.com/user-attachments/assets/5d8f402f-b928-4493-a43a-2e719fda21b2" />

- 
## Quantile Regression (very roughly)
1. As per first attack, get $S(x',y')$ for lots of records not used train model (I.e. $(x',y') \in \mathcal{D}^{out}$

2. Now order them from low to high and produce a chart of cumulative probability  (effectively quantiles) 
    - i.e.  for any given value $S_f(x',y')$ what proportion of the records had a value less than that.
    - so this goes from 0 to 1.0

3. Now train a model $\mathcal{Q}$ on all these pairs of $ (   (x','y'), quantile(S(x',y'))    
  i.e. train it to predict the solid redline in the picture: $\mathcal{Q} (x,y) \approx quantile(S_f(x,y))$
 
4. Given the record we want to test $(x,y)$   
- predict $\mathcal{Q}(x,y)$ 
- do a lookup-up of the quartile value for the test statistic $S_f(x,y)$ (the height of the red line at that point)  - if $\mathcal{Q}(x,y)   >> quartile(S_f(x,y)$ (the prediction is _above_ the red line) the record is likely to be in the training set
 -  if $\mathcal{Q}(x,y)   << quartile(S_f(x,y)$ (the prediction is _below_ the red line) the record is not likely to be in the training set
 -  

<img width="912" height="368" alt="Image" src="https://github.com/user-attachments/assets/bfc58af1-db29-4f20-9f7e-83978031b4e4" />

**Comment by jim-smith** [comment]
@shamykyzer @ssrhaso can you read through the above and let me know if it makes sense?

**Comment by shamykyzer** [comment]
> @shamykyzer @ssrhaso can you read through the above and let me know if it makes sense?

Hello Jim,

Will do, I'll look at it later tonight when I'm back home.

**Comment by ssrhaso** [comment]
Thanks for the writeup @jim-smith , really helpful, the progression from loss-based to shadow-model to quantile regression makes sense to me and as I was reading through the paper you sent I am also understanding this in more depth.

One thing I wanted to clarify before starting on an implementation: the sketch describes training Q to predict the quantile rank of an observed score (0 -> 1), whereas the paper trains Q to predict the score threshold at a fixed quantile level using pinball loss - where it predicts member if score > q_alpha(x). The paper's version gives the FPR calibration guarantee directly.

Is the intent to follow the paper's formulation, or the variant you shared? I am trying to understand if this is a SACRO-ML specific. 

Thank you!

**Comment by jim-smith** [comment]
Hi @ssrhaso Glad you found it helpful.
their formulation is right - I was trying to convey the broad idea rather than the specifics :)

and there is at least one version of their code in the links above.

**Comment by shamykyzer** [comment]
## Summary

Hello @jim-smith!

This PR implements the Quantile Regression Membership Inference Attack (QR-MIA) from Bertran et al. (NeurIPS 2023) as a new `QMIAAttack` class in `sacroml.attacks`. Unlike LiRA, QR-MIA requires **no shadow models**. It trains a single `HistGradientBoostingRegressor` with quantile loss to estimate per-sample membership thresholds. This makes it dramatically faster while delivering superior TPR at controlled FPR levels.

<table>
<tr>
<th>Implementation Architecture</th>
<th>Execution Flow</th>
</tr>
<tr>
<td><img src="https://github.com/user-attachments/assets/c62ea646-20a1-4b37-82d4-3f20d2bd7835" width="450"></td>
<td><img src="https://github.com/user-attachments/assets/da22e929-bb00-4763-8019-736f5d69ea13" width="450"></td>
</tr>
</table>

## Core implementation

- **`sacroml/attacks/qmia_attack.py`**: `QMIAAttack` class subclassing `Attack`. Pipeline: compute hinge scores, train quantile regressor, predict thresholds, convert margins to membership probabilities, report metrics.
- **`sacroml/attacks/utils.py`**: Shared utility functions: `logit`, `extract_true_label_probs`, `qmia_hinge_score` (vectorized, any C >= 2), `membership_labels`, `margins_to_two_column_probs`.

## Key design decisions

1. **`HistGradientBoostingRegressor` over `GradientBoostingRegressor`**: HGBT is up to 70x faster on large datasets with equivalent AUC. No external dependencies needed.

2. **No CatBoost dependency**: The original paper uses CatBoost, but sklearn's HGBT matches attack quality without adding an external dependency. scikit-learn is already a core dependency of SACRO-ML.

3. **Vectorized hinge score**: `qmia_hinge_score` computes `logit(p_y) - max_{y'!=y} logit(p_{y'})` entirely in NumPy. Works for any number of classes (C >= 2).

4. **Metrics adapter**: `margins_to_two_column_probs` converts raw score margins to the `[p_out, p_in]` two-column format expected by `sacroml.metrics.get_metrics()`.

## No new dependencies

The entire implementation uses `sklearn.ensemble.HistGradientBoostingRegressor`, which ships with scikit-learn, already a core SACRO-ML dependency.
